ledger/narwhal/compact-batch-header/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![forbid(unsafe_code)]
+#![warn(clippy::cast_possible_truncation)]
+
+mod bytes;
+mod serialize;
+mod string;
+mod to_id;
+
+use console::{
+    account::{Address, PrivateKey, Signature},
+    prelude::*,
+    types::Field,
+};
+use indexmap::{IndexMap, IndexSet};
+use narwhal_batch_header::BatchHeader;
+use narwhal_transmission_id::{TransmissionID, TransmissionType};
+
+/// Returns `true` if the transmission id indexes are properly formatted.
+fn sanity_check_indexes(transmission_ids_map: &IndexSet<(TransmissionType, u32)>) -> bool {
+    // Track the indexes for each transmission type.
+    let mut indices = IndexMap::new();
+
+    // Ensure that the index increments properly for each transmission type.
+    for (transmission_type, index) in transmission_ids_map {
+        let current_index = indices.entry(transmission_type).or_insert(0);
+        if *index != *current_index {
+            return false;
+        }
+        *current_index = current_index.saturating_add(1);
+    }
+
+    true
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct CompactBatchHeader<N: Network> {
+    /// The batch ID, defined as the hash of the round number, timestamp, transmission IDs, and previous batch certificate IDs.
+    batch_id: Field<N>,
+    /// The author of the batch.
+    author: Address<N>,
+    /// The round number.
+    round: u64,
+    /// The timestamp.
+    timestamp: i64,
+    /// The set of index references for the `transmission IDs`.
+    transmission_ids_map: IndexSet<(TransmissionType, u32)>,
+    /// The batch certificate IDs of the previous round.
+    previous_certificate_ids: IndexSet<Field<N>>,
```

Comment:
ditto, though at least here I can see the potential for things being queried for